runtime/lua/vim/lsp/buf.lua

```diff
@@ -37,24 +71,43 @@ function M.declaration()
   request('textDocument/declaration', params)
 end
 
-function M.definition()
-  local params = util.make_position_params()
-  request('textDocument/definition', params)
+--[[
+
+package.loaded['vim.lsp.actions'] = nil
+Location = require('vim.lsp.actions').Location
+
+vim.cmd "nnoremap <space>gd :lua vim.lsp.buf.definition {callbacks = { Location.jump_first, Location.highlight }}<CR>"
+vim.cmd "nnoremap <space>pd :lua vim.lsp.buf.definition {callbacks = Location.preview}<CR>"
+
+-- Builtin jump & highlight.
+vim.lsp.buf.definition { callbacks = { Location.jump, Location.highlight.with } } }
+
+-- Builtin jump. Highlight with override parameters
+vim.lsp.buf.definition { callbacks = { Location.jump, Location.highlight.with { timeout = 100, higroup = 'Error' } } }
+
+--]]
+
+--- Request textDocument/definition for current buffer.
+--
+-- <pre>
+-- nnoremap <space>gd <cmd>lua vim.lsp.buf.definition { callbacks = { Location.jump_first, Location.highlight } }<CR>
+-- nnoremap <space>pd <cmd>lua vim.lsp.buf.definition { callbacks = Location.preview }<CR>
+-- </pre>
+--
+function M.definition(args)
+  request('textDocument/definition', util.make_position_params(), args)
 end
 
-function M.type_definition()
-  local params = util.make_position_params()
-  request('textDocument/typeDefinition', params)
+function M.type_definition(args)
+  request('textDocument/typeDefinition', util.make_position_params(), args)
 end
 
-function M.implementation()
-  local params = util.make_position_params()
-  request('textDocument/implementation', params)
+function M.implementation(args)
+  request('textDocument/implementation', util.make_position_params(), args)
 end
 
-function M.signature_help()
-  local params = util.make_position_params()
-  request('textDocument/signatureHelp', params)
+function M.signature_help(args)
+  request('textDocument/signatureHelp', util.make_position_params(), args)
```

Comment:
Refactor the rest of the file in the same way... might be hard / break BWC (which is potentially fine. Could define new functions maybe)